boa_engine/src/builtins/uri/mod.rs

```diff
@@ -0,0 +1,564 @@
+                    if code_units[k] != 0x0025 {
+                        context
+                            .throw_uri_error("escape characters must be preceded with a % sign")?;
+                    }
+
+                    // c. If the code units at index (k + 1) and (k + 2) within string do not represent hexadecimal digits, throw a URIError exception.
+                    // d. Let B be the 8-bit value represented by the two hexadecimal digits at index (k + 1) and (k + 2).
+                    let b = decode_hex_byte(code_units[k + 1], code_units[k + 2], context)?;
+
+                    // e. Set k to k + 2.
+                    k += 2;
+
+                    // f. Append B to Octets.
+                    octets.push(b);
+
+                    // g. Set j to j + 1.
+                }
+
+                // 6. Assert: The length of Octets is n.
+                assert_eq!(octets.len(), n);
+
+                // 7. If Octets does not contain a valid UTF-8 encoding of a Unicode code point, throw a URIError exception.
+                match String::from_utf8(octets) {
+                    Err(_) => {
+                        return Err(context.construct_uri_error("invalid UTF-8 encoding found"))
+                    }
+                    Ok(v) => {
+                        // 8. Let V be the code point obtained by applying the UTF-8 transformation to Octets, that is, from a List of octets into a 21-bit value.
+
+                        // 9. Let S be UTF16EncodeCodePoint(V).
+                        // utf16_encode_codepoint(v)
+                        v.encode_utf16().collect::<Vec<_>>()
+                    }
+                }
+            }
+        };
+
+        // e. Set R to the string-concatenation of R and S.
+        r.extend_from_slice(&s);
+
+        // f. Set k to k + 1.
+        k += 1;
+    }
+}
+
+/// Decodes a byte from two unicode code units.
+fn decode_hex_byte(high: u16, low: u16, context: &mut Context) -> JsResult<u8> {
+    match (u8::try_from(high), u8::try_from(low)) {
+        (Ok(high), Ok(low)) => {
+            let high = if (b'0'..=b'9').contains(&high) {
+                high - b'0'
+            } else if (b'A'..=b'F').contains(&high) {
+                high - b'A' + 0x0A
+            } else if (b'a'..=b'f').contains(&high) {
+                high - b'a' + 0x0a
+            } else {
+                return Err(context.construct_uri_error("invalid hexadecimal digit found"));
+            };
+
+            let low = if (b'0'..=b'9').contains(&low) {
+                low - b'0'
+            } else if (b'A'..=b'F').contains(&low) {
+                low - b'A' + 0x0A
+            } else if (b'a'..=b'f').contains(&low) {
+                low - b'a' + 0x0a
+            } else {
+                return Err(context.construct_uri_error("invalid hexadecimal digit found"));
+            };
+
+            Ok((high << 4) + low)
+        }
+        _ => Err(context.construct_uri_error("invalid hexadecimal digit found")),
+    }
```

Comment:
This would simplify a lot if we casted `high` and `low` to `u32`, then tried to parse a `char` for each of them. That would allow us to call `char::to_digit` directly.

Reply:
I pushed a commit changing this to use `char::to_digit`. The generated asm looks better too as far as I can tell.